### PR TITLE
Expose offsets_for_times consumer method. closes #224

### LINKED ISSUE #224: Unable to start from beginning of topic using on_assign hook
I am unable to recreate the behavior of the kafka-console-consumer.sh script with my python code. I am using the broker-stored offsets (new style API).

I have a topic `io` which I am trying to read from the beginning with python.  This topic already has data in it (generated by a confluent_kafka Producer).

When I run the python consumer I see the consumer but it does not have any partition information:

```
$ ./bin/kafka-consumer-groups.sh --bootstrap kafka-0.io:9092 --list
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).

console-consumer-606
etl-3
console-consumer-66817
$ ./bin/kafka-consumer-groups.sh --bootstrap-server kafka-0.io:9092 --describe --group io-etl-3
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).


TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
$
```

The consumer code looks like the references:

```
$ cat kafka-to-postgres.py
import os
import requests
import time
import psycopg2
import json

from confluent_kafka import Consumer, KafkaError

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from models import Device

#### SQLALCHEMY CONNECTION ###
print "initializing db connection"
engine = create_engine( "postgresql+psycopg2://" +  os.environ['POSTGRES_USER'] + ":" + os.environ['POSTGRES_PW'] + "@" + os.environ['POSTGRES_HOST'] + ":" + os.environ["POSTGRES_SERVICE_PORT_PSQL"] + "/" + os.environ['POSTGRES_DB'] )

s = sessionmaker(bind=engine)
session = s()

#### KAKFA CONNECTTION ####
print "initializing kafka connection"
c = Consumer({
    'bootstrap.servers': os.environ['KAFKA_URL'] ,
    'group.id':          os.environ['KAFKA_CONSUMER_GROUP'],
    'auto.offset.reset': 'smallest',
})

def on_assign (c, ps):
  print("on_assign!!!")
  for p in ps:
    import pdb; pdb.set_trace()
    print("on_assigning")
    p.offset=-10
    c.assign(ps)
c.subscribe(['io'], on_assign=on_assign)

print "entering run loop"
running = True
while running:
    print "going to poll"
    msg = c.poll(timeout=1)
    if msg and not msg.error():
       # snip
    elif msg and msg.error().code() != KafkaError._PARTITION_EOF:
        print("error: %s"  % msg.error().str())
        running = False
    elif msg and msg.error():
        print("got error: %s" % ( msg.error().str() ) )
    else:
        print("got message %s" % (msg) )
c.close()
print "exiting"
exit(1)
```

With this output:

```
# KAFKA_CONSUMER_GROUP=io-etl-3 python kafka-to-postgres.py
initializing db connection
initializing kafka connection
entering run loop
going to poll
on_assign!!!
> /kafka-to-postgres.py(33)on_assign()
-> print("on_assigning")
(Pdb) l
 28
 29  	def on_assign (c, ps):
 30  	  print("on_assign!!!")
 31  	  for p in ps:
 32  	    import pdb; pdb.set_trace()
 33  ->	    print("on_assigning")
 34  	    p.offset=-10
 35  	    c.assign(ps)
 36  	c.subscribe(['io'], on_assign=on_assign)
 37
 38  	print "entering run loop"
(Pdb) p ps
[TopicPartition{topic=io,partition=0,offset=-1001,error=None}, TopicPartition{topic=io,partition=1,offset=-1001,error=None}]
(Pdb) p p
TopicPartition{topic=io,partition=0,offset=-1001,error=None}
(Pdb)
TopicPartition{topic=io,partition=0,offset=-1001,error=None}
(Pdb) p p.offset
-1001L
(Pdb) c
on_assigning
> /kafka-to-postgres.py(32)on_assign()
-> import pdb; pdb.set_trace()
(Pdb) c
on_assigning
got message None
going to poll
got message None
going to poll
got message None
going to poll
```

so no data comes through. I have tried setting the partition offset to `0`, `-10`. The default value provided to the `on_assign` callback is an offset of `-1001L`, if that makes any difference.

Compare to the output from invoking the kafka console consumer:

```
$ ./bin/kafka-console-consumer.sh --bootstrap-server kafka-0.io:9092 --topic io --from-beginning --timeout-ms=60000
# DATA DELUGE
```

and the offset inspector showing:

```
$ ./bin/kafka-consumer-groups.sh --bootstrap-server kafka-0:9092 --describe --group console-consumer-72507
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).

Consumer group 'console-consumer-72507' has no active members.

TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
io                          1          2393            2393            0          -                                                 -                              -
io                          0          2393            2393            0          -                                                 -
```

Why isn't the on_assign hook making any difference? I should see some kind of data coming through.

### PR DESCRIPTION
**Pull request by johnistan**
Expose offsets_for_times consumer method. closes #224

### CODE REVIEW

**Comment by johnistan** [comment]
I have tested this on an actual kafka instance, but there doesn't appear to be integration tests. So i have only checked in the basic smoke test.

**Comment by johnistan** [comment]
failing on untouched pep8 issues
```
./confluent_kafka/avro/cached_schema_registry_client.py:171:13: E722 do not use bare except'
./confluent_kafka/avro/cached_schema_registry_client.py:207:13: E722 do not use bare except'
./confluent_kafka/avro/cached_schema_registry_client.py:272:9: E722 do not use bare except'
./confluent_kafka/avro/serializer/message_serializer.py:45:1: E722 do not use bare except'
./confluent_kafka/avro/serializer/message_serializer.py:161:9: E722 do not use bare except'
./confluent_kafka/avro/serializer/message_serializer.py:183:13: E722 do not use bare except'
./examples/consumer.py:62:9: E722 do not use bare except'
./tests/test_threads.py:8:1: E722 do not use bare except'
./tests/avro/mock_registry.py:123:9: E722 do not use bare except'
```

**Comment by johnistan** [comment]
Other then the flake8 fails. anything this needs?

**Comment by johnistan** [comment]
bump. happy to respond to any comments.

File: confluent_kafka/src/Consumer.c [hunk 1]
```diff
@@ -936,6 +986,24 @@ static PyMethodDef Consumer_methods[] = {
...
+          ".. py:function:: offsets_for_times(partitions, [timeout=None])\n"
+          "\n"
+          " offsets_for_times looks up offsets by timestamp for the given partitions.\n"
+          "\n"
+          " The returned offsets for each partition is the earliest offset whose\n"
+          " timestamp is greater than or equal to the given timestamp in the\n"
+          " corresponding partition.\n"
+          "\n"
+          "  :param list(TopicPartition) partitions: topic+partitions with timestamps in the TopicPartition.offset field."
+          "  :param float timeout: Request timeout (when cached=False).\n"
```
Comment (edenhill) [inline confluent_kafka/src/Consumer.c]: remove cache mention

**Comment by edenhill** [comment]
Thank you for this, almost there!

**Comment by johnistan** [comment]
After some more testing I think i have found an issue with calling `offsets_for_time` with no timeout. The call just hangs and never returns. I need to dig in more but have you seen behavior like this before from the underlying C function?

**Comment by johnistan** [comment]
Found some more issues. The first call from a newly created consumer will fail like so:

```    
offsets = c.offsets_for_times(topic_partitions_to_search)
cimpl.KafkaException: KafkaError{code=LEADER_NOT_AVAILABLE,val=5,str="Failed to get offsets: Broker: Leader not available"}
```

I suspect the same issue a #196 and underlying https://github.com/edenhill/librdkafka/issues/1373

Adding a timeout causes it to work.

**Comment by johnistan** [comment]
It could be that my code has an issue but I think there are issues with the underlying C implementation. Without a formal integration setup that spins up a fresh kafka, sharing a reproducible example is a pain but here is a simple script:

This script shows two issues. First calling offset_for_times without a timeout or a -1 timeout, returns a copy of the TopicPartition objects you passed in. Second, calling `offset_for_times` without a timeout causes any subsequent `close()` on the consumer to hang.

```
import confluent_kafka
import uuid

topic_name = 'test-offsets'

conf = {
    'bootstrap.servers': 'localhost:9092',
    'api.version.request': True
}

# Create producer
p = confluent_kafka.Producer(**conf)

for i in range(1000):
    p.produce(topic_name, 'foo', timestamp=123456789000)
p.flush()

c_conf = {
    'bootstrap.servers': 'localhost:9092',
    'group.id': uuid.uuid1(),
    'session.timeout.ms': 6000,
    'api.version.request': True
}
c = confluent_kafka.Consumer(**c_conf)

topic_partitions_to_search = list(map(lambda p: confluent_kafka.TopicPartition(topic_name, p, 123456788000), range(0, 3)))
print("topic_partitions_to_search: %s" % topic_partitions_to_search)
# hack to make sure consumer has topic/broker metadata
c.subscribe([topic_name])
msg = c.poll()

offsets_with_timeout = c.offsets_for_times(topic_partitions_to_search, timeout=1.0)
offsets_with_neg_1_timeout = c.offsets_for_times(topic_partitions_to_search, timeout=-1)
offsets_without_timeout = c.offsets_for_times(topic_partitions_to_search)

print("offsets_with_timeout: %s" % offsets_with_timeout)
print("offsets_with_neg_1_timeout: %s" % offsets_with_neg_1_timeout)
print("offsets_without_timeout: %s" % offsets_without_timeout)

print("Right before close")
c.close()
print("Right after close")
```

Outputs:

```
py27 runtests: commands[1] | python examples/test.py
topic_partitions_to_search: [TopicPartition{topic=test-offsets,partition=0,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=123456788000,error=None}]
offsets_with_timeout: [TopicPartition{topic=test-offsets,partition=0,offset=8557,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=8179,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=9269,error=None}]
offsets_with_neg_1_timeout: [TopicPartition{topic=test-offsets,partition=0,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=123456788000,error=None}]
offsets_without_timeout: [TopicPartition{topic=test-offsets,partition=0,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=123456788000,error=None}]
Right before close
^CERROR: KEYBOARDINTERRUPT
C^CERROR: keyboardinterrupt
```

**Comment by johnistan** [comment]
Running locally with these changes to librdkafka: https://github.com/edenhill/librdkafka/pull/1525 i get good results:

```
py27 runtests: commands[1] | python examples/test.py
topic_partitions_to_search: [TopicPartition{topic=test-offsets,partition=0,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=123456788000,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=123456788000,error=None}]
offsets_with_timeout: [TopicPartition{topic=test-offsets,partition=0,offset=9888,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=9406,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=10711,error=None}]
offsets_with_neg_1_timeout: [TopicPartition{topic=test-offsets,partition=0,offset=9888,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=9406,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=10711,error=None}]
offsets_without_timeout: [TopicPartition{topic=test-offsets,partition=0,offset=9888,error=None}, TopicPartition{topic=test-offsets,partition=1,offset=9406,error=None}, TopicPartition{topic=test-offsets,partition=2,offset=10711,error=None}]
Right before close
Right after close
__________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________
  py27: commands succeeded
  congratulations :)
```

**Comment by matthew-d-jones** [comment]
@edenhill I think this should have closed #225 and not #224.

**Comment by DestroyerAlpha** [comment]
Is there proper documentation for this?

**Comment by matthew-d-jones** [comment]
@DestroyerAlpha https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Consumer.offsets_for_times

**Pull request merged**